### PR TITLE
adding ObjectiveLimit attribute

### PR DESCRIPTION
**Pull request by matbesancon**
This MR introduces an attribute to set an objective limit, after which the solver can stop.
This can be useful e.g. for separation routines, when one requires at least a certain objective value to be reached.

Setting this objective bound as a constraint in a feasibility formulation is often ill-advised and can lead to poor solver behaviour, most MIP solvers provide a dedicated parameter instead:
- Gurobi: https://www.gurobi.com/documentation/current/refman/bestobjstop.html#parameter:BestObjStop
- SCIP: `limits/objectivestop` is coming in the next major release
- HiGHS: objective_bound https://ergo-code.github.io/HiGHS/stable/options/definitions/#objective_bound

CPLEX seems to have [this](https://www.ibm.com/docs/en/icos/20.1.0?topic=parameters-lower-objective-value-limit) but it seems to be for LPs only?
Monotonically improving iterative solvers could also implement the attribute

### CODE REVIEW

File: src/attributes.jl [hunk 1]
```diff
@@ -850,6 +850,18 @@ struct TimeLimitSec <: AbstractOptimizerAttribute end
 
 attribute_value_type(::TimeLimitSec) = Union{Nothing,Float64}
 
+"""
+    ObjectiveLimit()
+
```
Comment (odow) [inline src/attributes.jl:855]:
```suggestion
```

File: src/Test/test_attribute.jl [hunk 2]
```diff
@@ -190,6 +190,30 @@ function setup_test(
     return
 end
 
+function test_attribute_ObjectiveLimit(model::MOI.AbstractOptimizer, ::Config)
+    @requires MOI.supports(model, MOI.ObjectiveLimit())
+    # Get the current value to restore it at the end of the test
+    value = MOI.get(model, MOI.ObjectiveLimit())
+    MOI.set(model, MOI.ObjectiveLimit(), 0.0)
+    @test MOI.get(model, MOI.ObjectiveLimit()) == 0.0
+    MOI.set(model, MOI.ObjectiveLimit(), 1.0)
```
Comment (odow) [inline src/Test/test_attribute.jl:199]: We should explicitly test setting `nothing` here as well

File: src/attributes.jl [hunk 3]
```diff
@@ -850,6 +850,18 @@ struct TimeLimitSec <: AbstractOptimizerAttribute end
...
+
+An optimizer attribute for setting a limit on the objective value.
```
Comment (odow) [inline src/attributes.jl:857]:
```suggestion
An optimizer attribute for setting a limit on the objective value.

```

File: src/attributes.jl [hunk 4]
```diff
@@ -850,6 +850,18 @@ struct TimeLimitSec <: AbstractOptimizerAttribute end
...
+The solver may stop when the `ObjectiveValue` is better (lower for minimization, higher for maximization) than the `ObjectiveLimit`.
```
Comment (odow) [inline src/attributes.jl:858]:
```suggestion
The solver may stop when the [`ObjectiveValue`](@ref) is better (lower for
minimization, higher for maximization) than the `ObjectiveLimit`. If stopped,
the [`TerminationStatus`](@ref) should be `OBJECTIVE_LIMIT`.

```

File: src/attributes.jl [hunk 5]
```diff
@@ -850,6 +850,18 @@ struct TimeLimitSec <: AbstractOptimizerAttribute end
...
+When `set` to `nothing`, it removes the solver limit. The default value is `nothing`.
```
Comment (odow) [inline src/attributes.jl:859]:
```suggestion
When `set` to `nothing`, the limit reverts to the solver's default.

The default value is `nothing`.
```

File: src/attributes.jl [hunk 6]
```diff
@@ -850,6 +850,18 @@ struct TimeLimitSec <: AbstractOptimizerAttribute end
...
+"""
+struct ObjectiveLimit <: AbstractOptimizerAttribute end
+
+attribute_value_type(::ObjectiveLimit) = Union{Nothing,Float64}
```
Comment (odow) [inline src/attributes.jl:863]: An implicit assumption is that this applies only to scalar-valued objective functions, and that the objective is `Float64`. What if the optimizer is `BigFloat`, or `Rational{BigFloat}`? And what about vector-valued objectives?

File: src/precompile.jl [hunk 7]
```diff
@@ -79,6 +79,7 @@ function precompile_model(model, constraints)
         ObjectiveValue,
         Silent,
         TimeLimitSec,
+        ObjectiveBound,
```
Comment (odow) [inline src/precompile.jl:82]: Unrelated?

File: src/Test/test_attribute.jl [hunk 8]
```diff
@@ -190,6 +190,30 @@ function setup_test(
     return
 end
 
+function test_attribute_ObjectiveLimit(model::MOI.AbstractOptimizer, ::Config)
+    @requires MOI.supports(model, MOI.ObjectiveLimit())
+    # Get the current value to restore it at the end of the test
+    value = MOI.get(model, MOI.ObjectiveLimit())
+    MOI.set(model, MOI.ObjectiveLimit(), 0.0)
+    @test MOI.get(model, MOI.ObjectiveLimit()) == 0.0
```
Comment (odow) [inline src/Test/test_attribute.jl:198]:
```suggestion
    @test MOI.get(model, MOI.ObjectiveLimit()) == 0.0
    MOI.set(model, MOI.ObjectiveLimit(), nothing)
    @test MOI.get(model, MOI.ObjectiveLimit()) === nothing
```

File: src/attributes.jl [hunk 9]
```diff
@@ -850,6 +850,25 @@ struct TimeLimitSec <: AbstractOptimizerAttribute end
...
+"""
+    ObjectiveLimit()
+
+An optimizer attribute for setting a limit on the objective value.
+
+The provided limit must be a `Union{Float64,Nothing}`.
+
+When `set` to `nothing`, the limit reverts to the solver's default.
+
+The default value is `nothing`.
```
Comment (odow) [inline src/attributes.jl:862]:
```suggestion
The default value is `nothing`.

```

File: src/attributes.jl [hunk 10]
```diff
@@ -850,6 +850,25 @@ struct TimeLimitSec <: AbstractOptimizerAttribute end
...
+The solver may stop when the [`ObjectiveValue`](@ref) is better (lower for
+minimization, higher for maximization) than the `ObjectiveLimit`. If stopped,
+the [`TerminationStatus`](@ref) should be `OBJECTIVE_LIMIT`.
+
```
Comment (odow) [inline src/attributes.jl:866]:
```suggestion
```

File: src/precompile.jl (hunk 7 above)
Comment (odow) [inline src/precompile.jl:82]:
```suggestion
```

File: src/attributes.jl (hunk 6 above)
Comment (matbesancon) [reply src/attributes.jl:870]: vetor-valued objectives is another story I wouldn't touch at first. I would be okay upgrading this to:
```
Union{Nothing,Real}
```

**Comment by matbesancon** [comment]
good to go?

File: src/attributes.jl (hunk 6 above)
Comment (odow) [reply src/attributes.jl:870]: Maybe we should just remove this for now. It's only implemented for a subset of attributes, and the fallback is `Any`.

Comment (matbesancon) [reply src/attributes.jl:870]: Sure yes

File: src/attributes.jl [hunk 11]
```diff
@@ -850,6 +850,25 @@ struct TimeLimitSec <: AbstractOptimizerAttribute end
...
+The default value is `nothing`.
+
+The solver may stop when the [`ObjectiveValue`](@ref) is better (lower for
+minimization, higher for maximization) than the `ObjectiveLimit`. If stopped,
+the [`TerminationStatus`](@ref) should be `OBJECTIVE_LIMIT`.
+"""
+struct ObjectiveLimit <: AbstractOptimizerAttribute end
+
+attribute_value_type(::ObjectiveLimit) = Union{Nothing,Float64}
```
Comment (matbesancon) [inline src/attributes.jl:870]:
```suggestion
```

File: src/attributes.jl [hunk 12]
```diff
@@ -850,6 +850,24 @@ struct TimeLimitSec <: AbstractOptimizerAttribute end
...
+
+The default value is `nothing`.
+
+The solver may stop when the [`ObjectiveValue`](@ref) is better (lower for
+minimization, higher for maximization) than the `ObjectiveLimit`. If stopped,
+the [`TerminationStatus`](@ref) should be `OBJECTIVE_LIMIT`.
+"""
+struct ObjectiveLimit <: AbstractOptimizerAttribute end
+
+
```
Comment (matbesancon) [inline src/attributes.jl:870]:
```suggestion
```

**Review by odow — APPROVED** [review]
I'm okay with this, but it probably needs @blegat to take a look.

**Comment by odow** [comment]
Bump @blegat

**Comment by odow** [comment]
Bump again @blegat

**Comment by matbesancon** [comment]
ping @joaquimg too in case

File: src/attributes.jl [hunk 13]
```diff
@@ -850,6 +850,23 @@ struct TimeLimitSec <: AbstractOptimizerAttribute end
 
 attribute_value_type(::TimeLimitSec) = Union{Nothing,Float64}
 
+"""
+    ObjectiveLimit()
+
+An optimizer attribute for setting a limit on the objective value.
+
+The provided limit must be a `Union{Float64,Nothing}`.
```
Comment (blegat) [inline src/attributes.jl:858]: It's not Float64. For a solver using BigFloat, it can be BigFloat. Let's just say it's Real

Comment (matbesancon) [reply src/attributes.jl]: done

**Review by blegat — APPROVED** [review]

**Comment by jajhall** [comment]
> * HiGHS: objective_bound https://ergo-code.github.io/HiGHS/stable/options/definitions/#objective_bound

This is applied to the dual objective in dual simplex, since it's a bound on the optimal primal objective value. So, it's not the option that should be set in the MIP context

HiGHS also has an `objective_target` option that was defined so that the primal simplex solver could stop if the target was reached. However this hasn't yet been implemented, and interpreting `objective_target` in the MIP context makes sense - and is easy to implement.

My only hesitation relates to the possibility of losing consistency with Gurobi. Its `BestObjStop` is just for the MIP context. Do you know whether Gurobi has a separate "target" parameter for continuous problems? My immediate search suggests not.

**Comment by matbesancon** [comment]
> My only hesitation relates to the possibility of losing consistency with Gurobi. Its BestObjStop is just for the MIP context. Do you know whether Gurobi has a separate "target" parameter for continuous problems? My immediate search suggests not.

not that I have seen no, I have seen it mostly applicable on MIP (on Gurobi and CPLEX)

**Comment by matbesancon** [comment]
the attribute defined here is independent of the type of problem solved. For an LP, using `objective_target` makes sense for LPs, I don't know whether this should force the use of the primal simplex or throw an error if not used together with the primal simplex?

Would that be okay to have the same parameter used for MIPs?

**Comment by jajhall** [comment]
> For an LP, using `objective_target` makes sense for LPs, I don't know whether this should force the use of the primal simplex or throw an error if not used together with the primal simplex?

Interesting thought. The default simplex strategy is to allow HiGHS to choose - and it always uses dual, since it's generally better and the implementation is better than the primal simplex implementation. I guess that it should throw an error if  `objective_target` has been set and the user specifies dual simplex, as it feeds back useful information.

**Comment by jajhall** [comment]
> Would that be okay to have the same parameter used for MIPs?

Looking at  [Gurobi status codes](https://www.gurobi.com/documentation/current/refman/optimization_status_codes.html) for something else yesterday, I followed the links for `USER_OBJ_LIMIT`, and both refer to MIP. So, it's safe to assume that LP target/bounds aren't available to users
